docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,354 @@
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions for the simulator
+
+The client is effectively an `IInflightHost`, so it any resource lifted into it registers the permissions it needs.
+This is implemented on a per-target basis to expose the permissions in a useful way.
+For the `sim` target this problem is (currently) much more difficult.
+The simulator is centralized and will have no knowledge of the future clients that will be using it.
+The simulator has no API to add new permissions after the simulator has started and the scope of that feels too large for this RFC.
+
+As an MVP, the simulator can be configured to ignore permissions via `wing.toml`.
+Even though this is unsafe and cause cause local/cloud differences, wing is intentionally not claiming ownership of the client's host environment.
+
+### Type generation with `@lift`
+
+Wing's compiler has type information available for use, but only before preflight execution occurs.
+This makes an API like the following not feasible to implement using logic similar to the existing `extern` or `@inflight`:
+
+```wing
+client.lift("bucket", bucket, ["put"]);
+```
+
+While executing preflight, the knowledge about the type of the `bucket` instance is lost, so generating the actual client code will have no way to reference it.
+
+This RFC proposes changing the existing `lift` syntax as a way to capture the necessary type information during preflight.
+An important feature of this syntax is that it's entirely statically analyzable, so anything we need to do with it can be done before-preflight.
+The existing `lift` syntax is currently intended to be used only inside inflights to explicitly lift preflight data into the defined scope:
+
+```wing
+let b1 = new cloud.Bucket();
+inflight () => {
+  let var b = b1;
+  lift { b: [put] } {
+    b.put("key", "value");
+  }
+}
+```
+
+With some changes, with syntax could also be used as a way to create a preflight container for lifted data.
+
+```wing
+let b = new cloud.Bucket();
+let lifts = @lift({ b: [put] });
+//  ^^^^^ is a std.LiftQualificationsContainer
+```
+
+This is very similar to the `lift` mechanism in `@wingcloud/framework`
+
+```ts
+const b = new cloud.Bucket(app, "Bucket");
+const lifts = lift({ bucket }).grant({ bucket: ["put"] });
+```
+
+In the [RFC for explicit lift qualifications](https://github.com/winglang/wing/blob/main/docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md), there is a mention of a `std.LiftQualificationsContainer`.
+
+If this data structure is augmented to also include type information, it could be used to generate the client during preflight.
```

Comment:
This feels like the right direction.

A good way to evaluate if we succeeded would be to implement `typescript.Client` purely in user land.